types/time.go

```diff
@@ -2980,76 +2980,157 @@ func minutesNumeric(t *CoreTime, input string, ctx map[string]int) (string, bool
 	return input[length:], true
 }
 
-const time12HourLen = len("hh:mm:ssAM")
+type parseState int32
 
-func time12Hour(t *CoreTime, input string, ctx map[string]int) (string, bool) {
-	// hh:mm:ss AM
-	if len(input) < time12HourLen {
-		return input, false
-	}
-	hour, succ := parseDigits(input, 2)
-	if !succ || hour > 12 || hour == 0 || input[2] != ':' {
-		return input, false
+const (
+	parseStateNormal    parseState = 1
+	parseStateFail      parseState = 2
+	parseStateEndOfLine parseState = 3
+)
+
+func parseSep(input string) (string, parseState) {
+	input = skipWhiteSpace(input)
+	if len(input) == 0 {
+		return input, parseStateEndOfLine
 	}
-	// 12:34:56 AM -> 00:34:56
-	if hour == 12 {
-		hour = 0
+	if input[0] == ':' {
+		return input[1:], parseStateNormal
 	}
+	return input, parseStateFail
+}
 
-	minute, succ := parseDigits(input[3:], 2)
-	if !succ || minute > 59 || input[5] != ':' {
-		return input, false
-	}
+func time12Hour(t *CoreTime, input string, ctx map[string]int) (string, bool) {
+	tryParse := func(input string) (string, parseState) {
+		state := parseStateNormal
+		// hh:mm:ss AM
+		/// Note that we should update `t` as soon as possible, or we
+		/// can not get correct result for incomplete input like "12:13"
+		/// that is shorter than "hh:mm:ss"
+		result := oneOrTwoDigitRegex.FindString(input) // 1..12
+		length := len(result)
+		hour, succ := parseDigits(input, length)
```

Comment:
Can these lines extract a function?  
I think extract a function like `func parseOneOrTwoDigit(input string) (num int, succ bool, input string)` can increase the readability of the code

Reply:
Actually finding the length by `oneOrTwoDigitRegex` is useless and costly, I combine them into one function in https://github.com/pingcap/tidb/pull/25389 and improve the performance by about 50%.